packages/jupyter-ai/jupyter_ai/chat_handlers/tools.py

```diff
@@ -0,0 +1,260 @@
+        file_paths = self.getToolFiles(self.tools_file_path)
+        tools = getTools(file_paths)
+        tool_node = ToolNode(tools)
+
+        # Bind tools to LLM
+        # Check if the LLM class takes tools else advise user accordingly.
+        # Can be extended to include temperature parameter
+        try:
+            self.model_with_tools = self.llm.__class__(
+                model_id=self.llm.model_id
+            ).bind_tools(tools)
+        except Exception as e:
+            self.reply(f"Not a chat model, cannot be used with tools. {e}")
+
+        # Initialize graph
+        agentic_workflow = StateGraph(MessagesState)
+        # Define the agent and tool nodes we will cycle between
+        agentic_workflow.add_node("agent", call_tool)
+        agentic_workflow.add_node("tools", tool_node)
+        # Add edges to the graph
+        agentic_workflow.add_edge("__start__", "agent")
+        agentic_workflow.add_conditional_edges("agent", self.conditional_continue)
+        agentic_workflow.add_edge("tools", "agent")
+        # Compile graph
+        app = agentic_workflow.compile()
+
+        # Run query
+        # res = ToolsChatHandler.toolChat(self, query) # For all Human and AI messages, if needed later
+        res = app.invoke({"messages": query})
+        return res["messages"][-1].content
+
+    async def process_message(self, message: HumanChatMessage):
+        args = self.parse_args(message)
+        if args is None:
+            return
+
+        if args.list:
+            tool_files = os.listdir(
+                os.path.join(Path.home(), ".jupyter/jupyter-ai/tools")
+            )
+            self.reply(f"The available tools files are: {tool_files}")
+            return
+        elif args.tools:
+            self.tools_file_path = os.path.join(
+                Path.home(), ".jupyter/jupyter-ai/tools", args.tools
+            )
+        else:
+            self.tools_file_path = os.path.join(
+                Path.home(), ".jupyter/jupyter-ai/tools"
+            )
+
+        query = " ".join(args.query)
+        if not query:
+            self.reply(f"{self.parser.format_usage()}", message)
+            return
+
+        self.get_llm_chain()
+
+        try:
+            with self.pending("Using LLM with tools ..."):
+                response = self.useLLMwithTools(query)
+                self.reply(response, message)
+        except Exception as e:
+            self.log.error(e)
+            response = """Sorry, tool usage failed.
+            Either (i) this LLM does not accept tools, (ii) there an error in
+            the custom tools file, (iii) you may also want to check the
+            location and name of the tools file, or (iv) you may need to install the
+            `langchain_<provider_name>` package. (v) Finally, check that you have
+            authorized access to the LLM."""
+            self.reply(response, message)
```

Comment:
Having a single error message can be really confusing to users since a user must explore each of the possibilities listed here to determine why `/tools` is not working on their instance. A better practice is to use custom exception classes to differentiate between runtime errors, then use those to provide specific error messages to tell the user exactly what went wrong and how to fix it.

For example, take the exception being caught in `getToolNames()`:

```py
   def getToolNames(self, tools_file_path):
        ...
        try:
            ...
        except FileNotFoundError as e:  # to do
            self.reply(f"Tools file not found at {tools_file_path}.")
```

Instead, this should re-raise a custom exception, such as:

```py

class ExceptionNoToolFile(Exception):
    pass

class ToolsChatHandler(...):
    ...
    def getToolNames(self, tools_file_path):
        ...
        try:
            ...
        except FileNotFoundError as e:
            raise ExceptionNoToolFile()
```

You can then repeat this for other errors that may be raised at runtime. Then, when the LLM is called, you can catch each of these custom errors to provide unique and more helpful error messages. For example:

```py
        try:
            with self.pending("Using LLM with tools ..."):
                response = self.useLLMwithTools(query)
                self.reply(response, message)
        except ExceptionNoToolFile:
            self.reply(f'No tool file found at {self.tools_file_path}')
        except ExceptionInvalidTool:
            self.reply('Failed to compile tool at ...')
        ...
```

You can also add fields to the custom exception classes to pass other info, like the tool file path. [See here for a reference implementation.](https://stackoverflow.com/a/1319675/12548458)

Reply:
Added error specific handling with custom Exception classes as follows:
```
class ExceptionNoToolsFile(Exception):
    """Missing tools file"""
    pass

class ExceptionModelDoesTakeTools(Exception):
    """Model is not a chat model that takes tools"""
    pass

class ExceptionModelNotAuthorized(Exception):
    """Authentication failed for model authorization"""
    pass
```
Associated `try` and `except` additions made for these errors.